src/wix/WixToolset.Core.Burn/Bundles/BurnReader.cs

```diff
@@ -142,6 +142,69 @@ public bool ExtractUXContainer(string outputDirectory, string tempDirectory)
             return true;
         }
 
+        /// <summary>
+        /// Extracts detached containers to the output directory.
+        /// </summary>
+        /// <param name="outputDirectory">Directory to write extracted files to.</param>
+        /// <param name="tempDirectory">Scratch directory.</param>
+        /// <param name="uxOutputDirectory">UX extraction folder. If null or empty, a UX folder will be created within tempDirectory</param>
+        /// <returns>True if successful, false otherwise</returns>
+        public bool ExtractDetachedContainers(string outputDirectory, string tempDirectory, string uxOutputDirectory)
+        {
+            if (string.IsNullOrEmpty(uxOutputDirectory))
+            {
+                uxOutputDirectory = Path.Combine(tempDirectory, "UX", "Final");
+            }
+
+            var manifestPath = Path.Combine(uxOutputDirectory, "manifest.xml");
+            if (!File.Exists(manifestPath))
+            {
+                var uxTempDirectory = Path.Combine(tempDirectory, "UX", "Temp");
+                if (!this.ExtractUXContainer(uxOutputDirectory, uxTempDirectory))
+                {
+                    return false;
+                }
+            }
+
+            Directory.CreateDirectory(tempDirectory);
+            Directory.CreateDirectory(outputDirectory);
+            var document = new XmlDocument();
+            document.Load(manifestPath);
+            var nsmgr = new XmlNamespaceManager(document.NameTable);
+            nsmgr.AddNamespace("burn", BurnCommon.BurnNamespace);
+            var detachedContainerNodes = document.SelectNodes("/burn:BurnManifest/burn:Container[not(@Attached = 'yes')]", nsmgr);
+
+            var exeDirectory = Path.GetDirectoryName(this.fileExe);
+            foreach (var node in detachedContainerNodes)
+            {
+                var containerElement = (XmlElement)node;
+                var containerName = containerElement.GetAttribute("FilePath");
+
+                var cabinetPath = Path.Combine(exeDirectory, containerName);
+                var extractDirectory = Path.Combine(tempDirectory, containerName);
+
+                Directory.CreateDirectory(extractDirectory);
+                var cabinet = new Cabinet(cabinetPath);
+                cabinet.Extract(extractDirectory);
+
+                var containerId = containerElement.GetAttribute("Id");
+                var containerPayloadNodes = document.SelectNodes($"/burn:BurnManifest/burn:Payload[@Packaging='embedded' and @Container='{containerId}']", nsmgr);
+                foreach (var payloadNode in containerPayloadNodes)
+                {
+                    var payloadElement = (XmlElement)payloadNode;
```

Comment:
Same as above, the cast can be avoided using the explicit type in the `foreach`.